Campaign summary statistics

tejasync opened the pull request:
| Q                                      | A
| -------------------------------------- | ---
| Branch?                                | campaign-summary <!-- see below -->
| Bug fix?                               | 
| New feature?                           | Yes
| Deprecations?                          | 
| BC breaks?                             | 
| Automated tests included?              | Yes
| Related user documentation PR URL      |  <!-- required for new features -->
| Related developer documentation PR URL |  <!-- required for developer-facing changes -->
| Issue(s) addressed                     | [#6651](https://github.com/mautic/mautic/pull/6651) <!-- prefix each issue number with "Fixes #", no need to create an issue if none exists, explain below instead -->

<!--
Additionally (see https://contribute.mautic.org/contributing-to-mautic/developer/code/pull-requests#step-5-work-on-your-pull-request):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against the "features" branch.
-->

**ToDo**

- [x] Write more functional tests to guarantee the charts and counts work as they should.
- [x] Fix the command as it seems it does not count contacts properly. There is a functional test that shows it.
- [x] Implement the Pending Count that has been added after this PR was submitted. It shows bad counts because of it after rebase.

<!--
Please write a short README for your feature/bugfix. This will help people understand your PR and what it aims to do.
-->
#### Description:
This PR fixes [#6651](https://github.com/mautic/mautic/pull/6651).

<!--
If you are fixing a bug and if there is no linked issue already, please provide steps to reproduce the issue here.
-->

Need run migration: `php bin/console doctrine:migrations:execute 20201120122846`

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Go to /s/config/edit and Campaign Settings.
3. Enable "Use summary statistics" and "Use date range"
4. Progress a contact through any campaign.
5. Go to the campaign /s/campaigns and view the Campaign Statistics chart. It should display the events exactly as you would expect.
6. View the Campaign Actions/Conditions/Decisions tabs (if there are any) and they should show the same results they always have as well.
7. Pick a date range with no contacts and hit apply. You'll see the campaign Actions/Conditions/Decisions/Contacts tabs disappear or contain no content. These tabs now respect the date range provided.

<!--
If you have any deprecations, list them here along with the new alternative.
If you have any backwards compatibility breaks, list them here.
-->

<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/9471"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

Commit:
Add Campaign Summary entities.

Commit:
Improve count accuracy.

Commit:
Count scheduled events when they occur.

Commit:
Add Scheduled event counts.

Commit:
Add alternative getCampaignLogCounts.

Commit:
Optionally use summary data for Campaign Statistics charts.

Commit:
Code cleanup.

Commit:
Add command to backfill historical campaign summary data.

Commit:
Improve query times.

Commit:
Minor cleanup.

Commit:
Append Executioner test.

Commit:
Correction to flag logic.

Commit:
Prevent counting with a false withTotalCount.

Before this the count query was still being ran, which in many cases is
a huge overhead when there are millions of leads.

Commit:
Allow extra counts to be skipped when loading contact grid.

We already have the lead count, and that extra lead count costs us many
seconds to query.  In addition supporting a dynamic orderBy allows us
to avoid a tmp table file sort on millions of rows, but we’ll get
effectively the same intended result, being the most recent leads added
to the campaign in 3ms vs 89s.

Commit:
Add an option to rebuild already summarized data.

After talking with John, I realized that if there is a SQL error at
high load, this would be one way to remedy the issue by cron.

Commit:
Support "Campaign by date range" feature

This merges in and refactors #6021 and includes some basic query
caching for those who enable APCu or similar at the ORM level.

Commit:
Log count corrections.

Commit:
Support pending event counts as well.

Commit:
Set coreParametersHelper.

Something recently dropped this.

heathdutton committed:
Remove extra result query.

Commit:
Match core lead-count for scheduled events.

Commit:
Summarize by Hour instead of by Day.

Per day was taking upwards of 1.2hrs per day in production while the
data is in motion, which can cause a table lock. By hour is slower
overall, but safer.

Commit:
Add hour being processed to stdout.

heathdutton committed:
Add better master/slave support.

Commit:
Prevent confusing percentages when filtering by date.

scottshipman committed:
fix log warning about missing $groupBy

Commit:
Summarize campaigns from one hour in arrears.

This will allow one to use this as an hourly cron job via --rebuild to
correct data should one have locks on mass inserts.

Commit:
Prevent by-reference notice.

scottshipman committed:
[ENG-890] save summaries at end of kickoff executioner or when count of summray entities grater than 100

scottshipman committed:
revert  back to private

heathdutton committed:
Remove unnecessary parameter.

escopecz committed:
CS Fixer applied

escopecz committed:
Removing dead code with Rector

escopecz committed:
Removing unused dependency

escopecz committed:
PHP 7, Mautic 3 syntax updates

escopecz committed:
Returning file permissions from 755 back to 644

escopecz committed:
Forgot to remove the unnecessary dependecy from config

escopecz committed:
reverting bad refactoring

escopecz committed:
Adding migration

escopecz committed:
Using DateTimeInterface for types

escopecz committed:
More readable HTML

escopecz committed:
Fixing array to string conversion error

escopecz committed:
Updating deprecated code

escopecz committed:
Fixing strict types

escopecz committed:
Adding basic SummarizeCommand functional test

tejasync committed:
fix failing test case

tejasync committed:
fix pending count in actions

tejasync committed:
fix dateTo and add test case for contact count

tejasync committed:
add counts and charts test cases

RCheesley commented:
Hi @ts-navghane and thanks for making this PR. Could you please use the post template to add details about what this PR is aiming to achieve, how people can reproduce the issue if it's a bug / how they can test the fix or test the new feature?

This will help us to triage your PR and also allows community members to understand how to test.

It also looks like some tests are failing, not sure if you intended to push all the other commits from 2018 in this PR? Let us know if you need any help!

tejasync committed:
fix failing test cases and phpstan fixes

hluchas commented:
@RCheesley please add @ts-navghane to verified contributors. TY!

RCheesley commented:
@hluchas they are already on the list, but this PR contains commits from @scottshipman who has not signed the agreement. Is that someone from Acquia/Mautic Inc?

hluchas commented:
> @hluchas they are already on the list, but this PR contains commits from @scottshipman who has not signed the agreement. Is that someone from Acquia/Mautic Inc?

No. cc @escopecz ?

kuzmany commented:
Why we clone https://github.com/mautic/mautic/pull/6651 instead push changes to original PR? 

Anyway @scottshipman should be from @heathdutton  team.

escopecz commented:
@kuzmany @ts-navghane does not have permissions to push to @heathdutton's PR.

hluchas committed:
Add async call to get campaign contacts (first page, TODO)

hluchas committed:
Move frontend logic to separated JS

hluchas committed:
Do not fetch campaign leads directly from view action

hluchas committed:
Use leads-container to insert changed content

hluchas committed:
Refactor code into one GET request to fill contact block

escopecz committed:
Using class constant instead of the string

escopecz commented on app/bundles/CampaignBundle/Controller/CampaignController.php:
```diff
@@ -673,14 +687,31 @@ protected function getViewArguments(array $args, $action)
-                $leadCount                = $this->getCampaignModel()->getRepository()->getCampaignLeadCount($entity->getId());
-                $campaignLogCounts        = $eventLogRepo->getCampaignLogCounts($entity->getId(), false, false, true);
+                $eventLogRepo             = $this->getDoctrine()->getManager()->getRepository(LeadEventLog::class);
                 $pendingCampaignLogCounts = $eventLogRepo->getCampaignLogCounts($entity->getId(), false, false);
```
I'm a bit afraid that this is loading the count form the statistical table instead of the summary table no matter if the `campaign_use_summary` config option is true or false. Can you think of a way how to get this pending count to the summary table as well?

tejasync replied:
I think I have a better solution to this, the pending counts are calculated based on total leads - log entries. We can use similar logic without use of `pendingCount` column in summary table.

escopecz replied:
It returns the same results on my test campaign. Good thinking!

tejasync committed:
fix pending counts, migration foreign table reference with prefix

hluchas committed:
TODO

hluchas committed:
Fix DOM target

hluchas committed:
Add doc

hluchas committed:
Refactor code

hluchas committed:
Add spinner when page is waiting for first load

hluchas committed:
Fix test

hluchas committed:
Keep contact page number in session

escopecz reviewed (dismissed):
The code looks good and it returns the same results as before with or without the config options turned on. 👍

hluchas committed:
Prevent JS XHR run on other page than campaign detail

tejasync committed:
Fixed broken logic. Fixed breaking test cases. Added more test cases.

tejasync committed:
Fixed hidden config error. Optimized import statments and return declaration type.

tejasync committed:
Fix typo

escopecz committed:
Optimizations:

1. No need to use the ready function as the Mautic.*OnLoad functions are loaded when the dom is ready anyway.
2. mQuery(containerId) is instantiated only once to save some code execution.
3. The code is encapsulated in its own function so the comment is not needed and the campaignOnLoad function is not getting any longer.

hluchas commented:
I ended with this result even if contact was correctly processed through condition `Has valid e-mail address` whith action updated the contact properly.

Reproduction steps.
1. Add contact to segment and use it in new campaign
2. 1. Use `Has valid email` condition
3. Use `Update contact action` with Yes/No added to the title with result of the condition
4. Click `Conditions` tab and you'll see this

![Snímek obrazovky 2021-01-08 v 13 48 16](https://user-images.githubusercontent.com/12815758/104017413-5e213900-51b8-11eb-8d3a-c042e72f3f7d.png)

Let me know when this will be fixed @ts-navghane ;-) TY!

hluchas reviewed (changes requested):
Does not work as expected.
Description above

tejasync committed:
Break down code into separate functionality for better readability.

tejasync committed:
Removed some unwanted code and debug statements.

tejasync committed:
Added logCountsProcessed column.

tejasync committed:
Updated code to insert log_counts_processed.

tejasync committed:
Summary model changes to update log_counts_processed.

tejasync committed:
Checked getCampaignLogCounts() for event_id.

tejasync committed:
Checked getCampaignLogCounts() for event_id.

tejasync committed:
Clean campaign_summary before rebuild to avoid duplicates.

tejasync committed:
Added test case for campaign pending count.

tejasync committed:
Added down migration.

tejasync committed:
Fixed failing test case and some improvements.

tejasync committed:
Fix index name typo.

tejasync committed:
Fixed migration id data type issue.

tejasync committed:
Extended PreUpAssertionMigration.

tejasync committed:
Extended PreUpAssertionMigration.

tejasync committed:
Added skip assertions for FK.

tejasync committed:
Fixed SQL insert on duplicate key update error.

tejasync committed:
Removed updating the id, on duplicate key update.

tejasync committed:
Removed updating the id, on duplicate key update.

tejasync committed:
Fixed SQL error.

tejasync committed:
Added debug logs and some improvements.

tejasync committed:
Updated log_counts_processed to be default null.

tejasync committed:
Updated log_counts_processed to be default null.

tejasync committed:
Changed approach to update log_counts_processed.

tejasync committed:
Changed approach to update log_counts_processed.

tejasync committed:
Updated migration to make log_counts_processed default null.

tejasync committed:
Updated migration to make log_counts_processed default null.

tejasync committed:
Updated migration to make log_counts_processed default null.

tejasync committed:
Added provision to print process required time.

tejasync committed:
Fixed log calculation  when saving campaign summary.

tejasync committed:
Changed approach to insert summaries.

tejasync committed:
Updated insert query and removed unused methods.

tejasync committed:
Added summary persist after every 100 log processed.

tejasync committed:
Fixed table prefix issue.

tejasync committed:
Fixed rebuilding of summary from current hour.

tejasync committed:
Fixed duplicate counts.

tejasync committed:
Removed delete all on rebuild summary.

RCheesley commented:
@ts-navghane we would like to consider this for the 4.1 release but there are conflicts and feedback to be addressed. Can you look to get these resolved so that we can consider it for that release and have it tested etc in time?
 
Thanks!

dadarya0 committed:
fix phhp stan and cs

dadarya0 committed:
fix test cases

dadarya0 committed:
fix test case

dadarya0 committed:
cs fixer

dadarya0 committed:
fix cs issue

dadarya0 committed:
remove  var

dadarya0 committed:
fix test cases

dadarya0 commented:
@RCheesley conflict resolved and failing test cases are fixed now it is ready to test.

kuzmany reviewed (dismissed):
Looks good.
Everything works like before

Command `mautic:campaigns:summarize` work as expected as well.

Nice improvement 👍

RCheesley commented:
@ts-navghane there is a conflict here to resolve - once that is done we just need another test and it should be good to go. Please also pull in changes so we can test with Gitpod :).

Thanks!

tejasync committed:
Fix conflicts redeclaration error

tejasync committed:
Fix PHPStan

tejasync committed:
Fix PHPStan, CS Fixes

tejasync committed:
Fix more PHPStan issues.

tejasync committed:
Generating PHPStan baseline to fix not so obvious errors

RCheesley commented:
@ts-navghane sorry we have a couple more conflicts to be addressed then we should be good to merge assuming that [this feedback](https://github.com/mautic/mautic/pull/9471#issuecomment-756740599) was addressed? I couldn't clearly see from the following activity if it was tested / addressed.

I've also added the historical contributor that was flagging up the CLA bot to our list so that should be addressed.

tejasync committed:
Merge remote-tracking branch 'origin/4.x' into campaign-summary

* origin/4.x:
  form conditions fields (#8794)
  4.0  to 4.x merge (#10628)
  Display labels for custom contact columns (#10623)

tejasync committed:
Create new phpstan baseline to eliminate obvious errors again

tejasync commented:
> @ts-navghane sorry we have a couple more conflicts to be addressed then we should be good to merge assuming that [this feedback](https://github.com/mautic/mautic/pull/9471#issuecomment-756740599) was addressed? I couldn't clearly see from the following activity if it was tested / addressed.
> 
> I've also added the historical contributor that was flagging up the CLA bot to our list so that should be addressed.

Hi @RCheesley,
Thank you for following up.
Yes, the said bug is addressed and resolved.
Attaching a screenshot for reference. ![Screenshot 2021-11-24 at 18 50 47](https://user-images.githubusercontent.com/54406786/143246058-3851001a-1163-4a0b-ae29-0bf5554c9f1e.png)

Steps I did:
1. Add 2 contacts, one with email and one without
2. Create relevant segment and keep this segment as the campaign source
2. Use `Has valid email address` condition
3. Add `Update contact` action (any update) in both Yes/No of above condition
4. Execute:
  4.1. `mautic:segments:update` - adds 2 contacts to the segment
  4.2. `mautic:campaigns:update` - adds 2 contacts to the campaign
  4.1. `mautic:campaigns:trigger` - execute events of the campaign
5. Observe the actions and conditions tab in the campaign details page

Thank you.

RCheesley committed:
Merge branch '4.x' into campaign-summary

RCheesley committed:
Merge branch '4.x' into campaign-summary
